test/exe/admin_response_test.cc

```diff
@@ -0,0 +1,334 @@
+  AdminResponseSharedPtr response = streamingResponse();
+  waitForHeaders(response);
+  response->cancel();
+  int chunk_calls = 0;
+  response->nextChunk([&chunk_calls](Buffer::Instance&, bool) { ++chunk_calls; });
+  EXPECT_TRUE(quitAndWait());
+  EXPECT_EQ(0, chunk_calls);
+}
+
+TEST_F(AdminStreamingTest, CancelBeforeAskingForChunk2) {
+  AdminResponseSharedPtr response = streamingResponse();
+  waitForHeaders(response);
+  blockMainThreadAndCancelResponseAfterResume(response);
+  int chunk_calls = 0;
+  response->nextChunk([&chunk_calls](Buffer::Instance&, bool) { ++chunk_calls; });
+  resume_.Notify();
+  EXPECT_TRUE(quitAndWait());
+  EXPECT_EQ(0, chunk_calls);
+}
+
+TEST_F(AdminStreamingTest, CancelAfterAskingForChunk) {
+  AdminResponseSharedPtr response = streamingResponse();
+  waitForHeaders(response);
+  blockMainThreadUntilResume("/ready", "GET");
+  int chunk_calls = 0;
+
+  // Cause the /streaming handler to pause while yielding the next chunk, to hit
+  // an early exit in MainCommonBase::requestNextChunk.
+  next_chunk_hook_ = [response]() { response->cancel(); };
+  response->nextChunk([&chunk_calls](Buffer::Instance&, bool) { ++chunk_calls; });
+  resume_.Notify();
+  EXPECT_TRUE(quitAndWait());
+  EXPECT_EQ(0, chunk_calls);
+}
+
+TEST_F(AdminStreamingTest, QuitBeforeHeaders) {
+  AdminResponseSharedPtr response = streamingResponse();
+  EXPECT_TRUE(quitAndWait());
+  ChunksBytes chunks_bytes = runStreamingRequest(response);
+  EXPECT_EQ(1, chunks_bytes.first);
+  EXPECT_EQ(0, chunks_bytes.second);
+}
+
+TEST_F(AdminStreamingTest, QuitDeleteRace) {
+  AdminResponseSharedPtr response = streamingResponse();
+  // Initiates a streaming quit on the main thread, but do not wait for it.
+  quitAndRequestHeaders();
+  response.reset(); // Races with the quitquitquit
+  EXPECT_TRUE(waitForEnvoyToExit());
+}
+
+TEST_F(AdminStreamingTest, QuitCancelRace) {
+  AdminResponseSharedPtr response = streamingResponse();
+  quitAndRequestHeaders();
+  response->cancel(); // Races with the quitquitquit
+  EXPECT_TRUE(waitForEnvoyToExit());
+}
+
+TEST_F(AdminStreamingTest, QuitBeforeCreatingResponse) {
+  // Initiates a streaming quit on the main thread, and wait for headers, which
+  // will trigger the termination of the event loop, and subsequent nulling of
+  // main_common_. However we can pause the test infrastructure after the quit
+  // takes hold leaving main_common_ in tact, to reproduce a potential race.
+  pause_after_run_ = true;
+  adminRequest("/quitquitquit", "POST");
+  pause_point_.WaitForNotification(); // run() finished, but main_common_ still exists.
+  AdminResponseSharedPtr response = streamingResponse();
+  ChunksBytes chunks_bytes = runStreamingRequest(response);
+  EXPECT_EQ(1, chunks_bytes.first);
+  EXPECT_EQ(0, chunks_bytes.second);
+  resume_.Notify();
+  EXPECT_TRUE(waitForEnvoyToExit());
+  response.reset();
+}
+
+TEST_F(AdminStreamingTest, QuitTerminateRace) {
```

Comment:
Aren't quit and terminate the same 'cause'? I think this test is quit/terminate racing with AdminResponse deletion?